.github/workflows/code_coverage.yml

```diff
@@ -0,0 +1,38 @@
+name: Code Coverage
+
+on:
+  pull_request:
+    types: [opened, synchronize]
+  push:
+    branches:
+      - main  # Update this with your default branch
+
+jobs:
+  code_coverage:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Set up Rust
+        uses: actions/setup-rust@v1
+
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Install code coverage dependencies
+        run: |
+          rustup component add llvm-tools-preview
+          cargo install cargo-tarpaulin
+
+      - name: Run tests and generate coverage report
+        run: cargo tarpaulin --all --ciserver github-actions --coveralls ${{ github.event.pull_request.head.ref }}
+
+      - name: Publish code coverage comment on PR
+        uses: mikepenz/action-github-comment@v2
+        with:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

Comment:
I was referring to [this](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idpermissions)

will need to give write permissions on the PR so that it's able to add a comment
```
permissions: 
   pull-requests: write 
```

Reply:
Done this part